ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,239 @@
+        # type: () -> List
+        """Serialize and update the buckets.
+
+        The current bucket is left in case any other spans are added.
+        """
+        serialized_buckets = []
+        serialized_bucket_keys = []
+        now_ns = time_ns()
+        for bucket_time_ns, bucket in self._buckets.items():
+            # TODO: have to be able to force these out on process shutdown
+            if bucket_time_ns > now_ns - self._bucket_size_ns:
+                # do not flush the current bucket
+                # DEV: is this actually required? Can a bucket not be reported twice?
+                #      What about spans `finished` with a custom end time (in the past)
+                #      after the bucket has been flushed?
+                continue
+            bucket_aggr_stats = []
+            serialized_bucket_keys.append(bucket_time_ns)
+
+            for aggr_key, stat_aggr in bucket.items():
+                name, service, resource, _type, http_status, synthetics = aggr_key
+                bucket = {
+                    "Name": name,
+                    "Resource": resource,
+                    "Synthetics": synthetics,
+                    "Hits": stat_aggr.hits,
+                    "TopLevelHits": stat_aggr.top_level_hits,
+                    "Duration": stat_aggr.duration,
+                    "Errors": stat_aggr.errors,
+                    "OkSummary": DDSketchProto.to_proto(stat_aggr.ok_distribution).SerializeToString(),
+                    "ErrorSummary": DDSketchProto.to_proto(stat_aggr.err_distribution).SerializeToString(),
+                }
+                if service:
+                    bucket["Service"] = service
+                if _type:
+                    bucket["Type"] = _type
+                bucket_aggr_stats.append(bucket)
+            serialized_buckets.append(
+                {
+                    "Start": bucket_time_ns,
+                    "Duration": self._bucket_size_ns,
+                    "Stats": bucket_aggr_stats,
+                }
+            )
+
+        # Clear out buckets that have been serialized
+        for key in serialized_bucket_keys:
+            del self._buckets[key]
+
+        return serialized_buckets
+
+    def periodic(self):
+        # type: (...) -> None
+        with self._lock:
+            serialized_stats = self._serialize_buckets()
+
+        raw_payload = {"Stats": serialized_stats}
+        hostname = get_hostname()
+        if hostname:
+            raw_payload["Hostname"] = hostname
+        if config.env:
+            raw_payload["Env"] = config.env
+        if config.version:
+            raw_payload["Version"] = config.version
+
+        payload = packb(raw_payload)
+        headers = {
+            "Datadog-Meta-Lang": "python",
+            "Datadog-Meta-Tracer-Version": ddtrace.__version__,
+            "Content-Type": "application/msgpack",
+        }
+        try:
+            self._connection = get_connection(self._agent_url, self._timeout)
```

Comment:
todo: retries?

should we have a different retry policy from traces?

can we remove the retry logic from traces when stats computation is enabled? it'll be safer to be missing traces if we know we can get stats out.